engines/alg/graphics.cpp

```diff
@@ -0,0 +1,200 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#include "common/array.h"
+#include "common/debug.h"
+#include "common/file.h"
+#include "common/path.h"
+#include "common/rect.h"
+
+#include "alg/alg.h"
+#include "alg/graphics.h"
+
+namespace Alg {
+
+Graphics::Surface *AlgGraphics::loadVgaBackground(const Common::Path &path, uint8 *palette) {
+	Common::File vgaFile;
+	if (!vgaFile.open(path)) {
+		error("AlgGraphics::loadVgaBackground(): Can't open background file '%s'", path.toString().c_str());
+	}
+	uint16 width = vgaFile.readUint16LE();
+	uint16 height = vgaFile.readUint16LE();
+	uint8 paletteEntries = vgaFile.readByte();
+	uint8 paletteStart = vgaFile.readByte();
+	assert(width >= 317 && width <= 320);
+	assert(height == 200);
+	assert(paletteStart == 0x10);
+	for (uint32 i = paletteStart * 3; i < (paletteStart + paletteEntries) * 3U; i += 3) {
+		palette[i] = vgaFile.readByte();
+		palette[i + 1] = vgaFile.readByte();
+		palette[i + 2] = vgaFile.readByte();
+	}
+	Graphics::Surface *surface = new Graphics::Surface();
+	surface->create(width, height, Graphics::PixelFormat::createFormatCLUT8());
+	uint8 *pixels = new uint8[width * height]();
+	vgaFile.read(pixels, width * height);
+	surface->setPixels(pixels);
+	Common::Rect backgroundRect = Common::Rect(0, 0, width, height);
+	surface->flipVertical(backgroundRect);
+	vgaFile.close();
+	return surface;
+}
+
+// for "normal" ani images
+Common::Array<Graphics::Surface *> *AlgGraphics::loadAniImage(const Common::Path &path, uint8 *palette) {
+	Common::Array<Graphics::Surface *> *images = new Common::Array<Graphics::Surface *>();
+	Common::File aniFile;
+	if (!aniFile.open(path)) {
+		error("AlgGraphics::loadAniImage(): Can't open image file '%s'", path.toString().c_str());
+	}
+	uint8 paletteEntries = aniFile.readByte();
+	uint8 paletteStart = aniFile.readByte();
+	for (uint32 i = paletteStart * 3; i < (paletteStart + paletteEntries) * 3U; i += 3) {
+		palette[i] = aniFile.readByte();
+		palette[i + 1] = aniFile.readByte();
+		palette[i + 2] = aniFile.readByte();
+	}
+	uint16 length, width, height;
+	while (aniFile.pos() < aniFile.size()) {
+		width = height = 0;
+		uint64 aniSectionOffset = aniFile.pos();
+		while (aniFile.pos() < aniFile.size()) {
+			length = aniFile.readUint16LE();
+			if (length == 0) {
+				break;
+			}
+			width = length;
+			height++;
+			aniFile.skip(2 + length);
+		}
+		if (width > 0) {
+			aniFile.seek(aniSectionOffset, SEEK_SET);
+			Graphics::Surface *aniImage = new Graphics::Surface();
+			aniImage->create(width, height, Graphics::PixelFormat::createFormatCLUT8());
+			for (uint16 y = 0; y < height; y++) {
+				aniFile.skip(4);
+				for (uint16 x = 0; x < width; x++) {
+					aniImage->setPixel(x, y, aniFile.readByte());
+				}
```

Comment:
```suggestion
				aniFile.read(aniImage->getBasePtr(0, y), width);
```

Something similar can also be done in other places that read in rows of pixels at a time.

Reply:
same as above, will look into it

Reply:
fixed